docs/Troubleshooting.md

```diff
@@ -0,0 +1,159 @@
+# Troubleshooting
+
+## Not large enough perf_event_mlock_kb
+```
+perf_event mmap failed: Operation not permitted
+```
+Profiler allocates 8kB perf_event buffer for each thread of the target process.
+Make sure `/proc/sys/kernel/perf_event_mlock_kb` value is large enough
+(more than `8 * threads`) when running under unprivileged user. Otherwise, the above message
+will be printed, and no native stack traces will be collected.
+
+## Operation not permitted
+```
+Failed to change credentials to match the target process: Operation not permitted
+```
+Due to limitation of HotSpot Dynamic Attach mechanism, the profiler must be run
+by exactly the same user (and group) as the owner of target JVM process.
+If profiler is run by a different user, it will try to automatically change
+current user and group. This will likely succeed for `root`, but not for
+other users, resulting in the above error.
+
+## Unable to establish communication with the target JVM
+```
+Could not start attach mechanism: No such file or directory
+```
+
+The profiler cannot establish communication with the target JVM through UNIX domain socket.
+
+Usually this happens in one of the following cases:
+1. Attach socket `/tmp/.java_pidNNN` has been deleted. It is a common
+   practice to clean `/tmp` automatically with some scheduled script.
+   Configure the cleanup software to exclude `.java_pid*` files from deletion.  
+   How to check: run `lsof -p PID | grep java_pid`  
+   If it lists a socket file, but the file does not exist, then this is exactly
+   the described problem.
+2. JVM is started with `-XX:+DisableAttachMechanism` option.
+3. `/tmp` directory of Java process is not physically the same directory
+   as `/tmp` of your shell, because Java is running in a container or in
+   `chroot` environment. `jattach` attempts to solve this automatically,
+   but it might lack the required permissions to do so.  
+   Check `strace build/jattach PID properties`
+4. JVM is busy and cannot reach a safepoint. For instance,
+   JVM is in the middle of long-running garbage collection.  
+   How to check: run `kill -3 PID`. Healthy JVM process should print
+   a thread dump and heap info in its console.
+
+## libasyncProfiler load failed
+```
+Target JVM failed to load libasyncProfiler.so
+```
+The connection with the target JVM has been established, but JVM is unable to load profiler shared library.
+Make sure the user of JVM process has permissions to access `libasyncProfiler.so` by exactly the same absolute path.
+For more information see [#78](https://github.com/async-profiler/async-profiler/issues/78).
+
+## No access to perf event or perf events unavailable
+```
+No access to perf events. Try --fdtransfer or --all-user option or 'sysctl kernel.perf_event_paranoid=1'
```

Comment:
Some of the error messages were altered since their first appearance. Could you please verify that all messages are up-to-date?

Reply:
Thanks Andrei, this was a very important catch from your side. Yes there are changes. I will correct them.

Reply:
Addressed in [e6282c8](https://github.com/async-profiler/async-profiler/pull/1029/commits/e6282c83bc0aa6007df310fd5e638e9d4d5ed6f1)